File: Data Structures & Algorithms/house-robber/submission-2.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        if len(nums) == 1:
            return nums[0]

        if len(nums) == 2:
            return max(nums)

        h0, h1 = 0, 0
        for i in range(len(nums)):
            temp = h1
            h1 = max(h1, h0 + nums[i])
            h0 = temp

        return h1

File: Data Structures & Algorithms/maximum-product-subarray/submission-2.kt
class Solution {
    fun maxProduct(nums: IntArray): Int {
        var min_so_far = nums[0]
        var max_so_far = nums[0]
        var res = max_so_far

        for (i in 1 until nums.size){
            var n = nums[i]
            var temp = maxOf(n, n * min_so_far, n*max_so_far)
            min_so_far = minOf(n, n * max_so_far, n *  min_so_far)
            max_so_far = temp
            res = maxOf(res, max_so_far)
        }
        return res
    }   
}
